cluster/network.go

```diff
@@ -209,3 +231,236 @@ func (c *Cluster) getNetworkPluginManifest(pluginConfig map[string]string) (stri
+			return fmt.Errorf("[network] Can't access KubeAPI port [%s] on Control Plane host: %s", KubeAPIPort, host.Address)
+		}
+		conn.Close()
+	}
+	return nil
+}
+
+func (c *Cluster) deployTCPPortListeners(ctx context.Context) error {
+	log.Infof(ctx, "[network] Deploying port listener containers")
+	var errgrp errgroup.Group
+
+	portList := []string{
+		KubeAPIPort,
+		EtcdPort1,
+		EtcdPort2,
+		ScedulerPort,
+		ControllerPort,
+		KubeletPort,
+		KubeProxyPort,
+	}
+	portBindingList := []nat.PortBinding{}
+	for _, portNumber := range portList {
+		rawPort := fmt.Sprintf("0.0.0.0:%s:1337/tcp", portNumber)
+		portMapping, _ := nat.ParsePortSpec(rawPort)
+		portBindingList = append(portBindingList, portMapping[0].Binding)
+	}
+
+	imageCfg := &container.Config{
+		Image: c.SystemImages[AplineImage],
+		Cmd: []string{
+			"nc",
+			"-kl",
+			"-p",
+			"1337",
+			"-e",
+			"echo",
+		},
+		ExposedPorts: nat.PortSet{
+			"1337/tcp": {},
+		},
+	}
+
+	hostCfg := &container.HostConfig{
+		PortBindings: nat.PortMap{
+			"1337/tcp": portBindingList,
+		},
+	}
+	hostList := []*hosts.Host{}
+	hostList = append(hostList, c.EtcdHosts...)
+	hostList = append(hostList, c.ControlPlaneHosts...)
+	hostList = append(hostList, c.WorkerHosts...)
+	// little trick to get a unique host list
+	uniqHosts := make(map[*hosts.Host]bool)
+	for _, host := range hostList {
+		uniqHosts[host] = true
+	}
+	for host := range uniqHosts {
+		runHost := host
+		errgrp.Go(func() error {
+			return docker.DoRunContainer(ctx, runHost.DClient, imageCfg, hostCfg, PortListenContainer, runHost.Address, "network")
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+	log.Infof(ctx, "[network] Port listener containers deployed successfully")
+	return nil
+}
+func (c *Cluster) removeTCPPortListeners(ctx context.Context) error {
+	log.Infof(ctx, "[network] Removing port listener containers")
+	var errgrp errgroup.Group
+
+	hostList := []*hosts.Host{}
+	hostList = append(hostList, c.EtcdHosts...)
+	hostList = append(hostList, c.ControlPlaneHosts...)
+	hostList = append(hostList, c.WorkerHosts...)
+
+	for _, host := range hostList {
+		runHost := host
+		errgrp.Go(func() error {
+			return docker.DoRemoveContainer(ctx, runHost.DClient, PortListenContainer, runHost.Address)
+		})
+	}
+	if err := errgrp.Wait(); err != nil {
+		return err
+	}
+	log.Infof(ctx, "[network] Port listener containers removed successfully")
+	return nil
+}
+
+func (c *Cluster) runServicePortChecks(ctx context.Context) error {
+	var errgrp errgroup.Group
+	// check etcd <-> etcd
+	etcdPortList := []string{
+		EtcdPort1,
+		EtcdPort2,
+	}
+	// one etcd host is a pass
+	if len(c.EtcdHosts) > 1 {
+		log.Infof(ctx, "[network] Running etcd <-> etcd port checks")
+		for _, host := range c.EtcdHosts {
+			runHost := host
+			errgrp.Go(func() error {
+				return checkPlaneTCPPortsFromHost(ctx, runHost, etcdPortList, c.EtcdHosts, c.SystemImages[AplineImage])
+			})
+		}
+	}
+	if err := errgrp.Wait(); err != nil {
```

Comment:
nit: this code would be hit only when len(c.EtcdHosts) > 1, so may be put it in inside the condition?